contracts/infrastructure/dapp/ParaswapFilter.sol

```diff
@@ -228,20 +279,59 @@ contract ParaswapFilter is BaseFilter {
         return success && abi.decode(res, (bool));
     }
 
-    function hasValidExchangeAdapters(IParaswap.Path[] memory _path) internal view returns (bool) {
+    function hasTradableTokens(address[] memory _tokens) internal view returns (bool) {
+        (bool success, bytes memory res) = tokenRegistry.staticcall(abi.encodeWithSignature("areTokensTradable(address[])", _tokens));
+        return success && abi.decode(res, (bool));
+    }
+
+    function hasValidPath(address _fromToken, IParaswap.Path[] memory _path) internal view returns (bool) {
         for (uint i = 0; i < _path.length; i++) {
             for (uint j = 0; j < _path[i].routes.length; j++) {
-                if(!adapters[_path[i].routes[j].exchange]) {
+                IParaswap.Route memory route = _path[i].routes[j];
+                if(route.targetExchange != address(0) && !targetExchanges[route.targetExchange]) {
                     return false;
                 }
+                if(route.exchange == uniV2Adapter) { 
+                    if(!hasValidUniV2Route(route.payload, uniFactory, uniInitCode)) {
+                        return false;
+                    }
+                } else if(route.exchange == sushiswapAdapter) { 
+                    if(!hasValidUniV2Route(route.payload, uniForkFactory1, uniForkInitCode1)) {
+                        return false;
+                    }
+                } else if(route.exchange == linkswapAdapter) { 
+                    if(!hasValidUniV2Route(route.payload, uniForkFactory2, uniForkInitCode2)) {
+                        return false;
+                    }
+                } else if(route.exchange == defiswapAdapter) { 
+                    if(!hasValidUniV2Route(route.payload, uniForkFactory3, uniForkInitCode3)) {
+                        return false;
+                    }
+                } else if(route.exchange == uniV1Adapter) { 
```

Comment:
Is this really needed? Is there any route that uses Uniswap V1?

Reply:
In my experience, 1inch still regularly routes trades via UniV1, especially for older tokens whose liquidity hasn't been fully migrated or simply when a temporary arbitrage makes a UniV1 trade more attractive. I assume the same is true for Paraswap. In any case, if this check is the last one to be done, it doesn't increase the gas cost of any trade.